### PR TITLE
Prepare Batch

### PR DESCRIPTION
**Pull request by behxyz**
Extend default prepare batch in Trainer

### Description
This PR implements the abstract `PrepareBatch` and concrete classes:
- PrepareBatchDefault
- PrepareBatchExtraInput
- PrepareBatchShuffle

### Status
Ready

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by behxyz** [comment]
I'll update this PR to make them all handlers.

**Comment by Nic-Ma** [comment]
Thanks @behxyz , handler is existing API in MONAI and ignite, and very straight forward.
Recommend to use it for most cases.

Thanks.

**Comment by behxyz** [comment]
The [unittests](https://github.com/Project-MONAI/MONAI/pull/1561/checks?check_run_id=1856851001#step:10:7762), involving randomness, are failing despite `set_determinism(0)`. I have tested on several machines and docker images (ubuntu and macos) with different version of pytorch and all worked fine. Any thought why it is failing here? @wyli @Nic-Ma

**Comment by wyli** [comment]
I think there's a bug with the global random state, before that's confirmed/fixed, please use the `setUp` and `tearDown` to set seed like https://github.com/Project-MONAI/MONAI/blob/db831ae096bec76371bb8c2e3db226d947a5ac06/tests/test_bounding_rectd.py#L28

**Comment by Nic-Ma** [comment]
Hi @wyli and @behxyz ,

I think maybe it's better to implement the `shuffle` feature with ignite handler instead of `batch_prepare`.
handler is more flexible and easier to investigate with other components, some draft code:
```py
from typing import TYPE_CHECKING, Dict, Optional
import torch
from monai.utils import exact_version, optional_import, ensure_tuple
from monai.config import KeysCollection
Events, _ = optional_import("ignite.engine", "0.4.2", exact_version, "Events")
if TYPE_CHECKING:
    from ignite.engine import Engine
else:
    Engine, _ = optional_import("ignite.engine", "0.4.2", exact_version, "Engine")


class PatchShuffler:
    """
    When every Dataset generates a list of image patches, DataLoader only shuffles at high level.
    This handler is used to totally shuffle all the patches for a iteration.
    Expected input data should be dictionary.

    """
    def __init__(self, keys: KeysCollection):
        self.keys = ensure_tuple(keys)

    def attach(self, engine: Engine) -> None:
        engine.add_event_handler(Events.ITERATION_STARTED, self)

    def __call__(self, engine: Engine) -> None:
        idx_rand = torch.randperm(length(engine.state.batch[self.keys[0]]))
        for k in self.keys:
            engine.state.batch[k] = engine.state.batch[k][idx_rand]
```

What do you think?

Thanks.

**Comment by behxyz** [comment]
I think implementing prepare_batch as handlers could be confusing. Since the workflow already have prepare_batch and anyways it will call the default one on the batches which convert them from dict to tuple. Now if we make a prepare batch handler as well. First of all, we should check and see if it is called before or after self.prepare_batch. Second of all, anyways it changes the input/output type of prepare_batch and we cannot reuse them in their original place. Third of all, if it is being called after self.prepare_batch, it will be tuple in/tuple out, and we can’t know what they are without checking the pre transforms.

**Comment by wyli** [comment]
thanks @behxyz, this PR proposes to shuffle within each batch, could you provide a paper reference for this approach for clarity?

my impression of ignite is that we should make use of the event handling system whenever possible instead of expanding `prepare_batch`. 

@vfdev-5 could you please comment on this if you have time? our goal here is to shuffle a training batch before the model.forward(batch): https://github.com/behxyz/MONAI/blob/prepare_batch/monai/engines/prepare_batch.py#L104-L109  which is done as a `prepare_batch` callable. But it could also be implemented as a handler https://github.com/Project-MONAI/MONAI/pull/1561#issuecomment-775963186 which one is preferred from ignite's perspective?

**Comment by vfdev-5** [comment]
@wyli thanks for mentioning! 

Yes, we provide a way to dynamically preprocess the input batch (please, see the note [here](https://pytorch.org/ignite/engine.html#ignite.engine.engine.Engine.run)) and this can be done with a configurable handler:
```python
trainer = ...

@trainer.on(Events.ITERATION_STARTED(custom_event_filtering))
def switch_batch(engine):
    engine.state.batch = preprocess_batch(engine.state.batch)
```
such that user could easily turn on/off this preprocessing or apply with a certain frequency...

Hope this could help to decide from your side.

**Comment by vfdev-5** [comment]
I'm curious in which case it is interesting to shuffle things ? AFAIK in most of the cases, loss functions average across the batch dimension and this shouldn't change gradients, right ?

**Comment by wyli** [comment]
> I'm curious in which case it is interesting to shuffle input batch ? AFAIK in most of the cases, loss functions average across the batch dimension and this shouldn't change gradients, right ?

I'd like to know more details as well :) @behxyz

**Comment by behxyz** [comment]
That's good point! I believe this is what inherited from one of our pathology use cases, [here](https://github.com/baidu-research/NCRF/blob/a8657eed4ea105ec80bb1a003a5afb464e1ce99f/wsi/bin/train.py#L57). I am not convinced that this shuffling has any effect either.

**Comment by vfdev-5** [comment]
There can be a practical use-case, for example, when using N devices and when a batch is sort of data unbalanced (e.g. [C1, C1, ..., C1, C2, C2]). Then we'd like to interleave data such that each process (and models with BN) sees more or less balanced data...

**Comment by wyli** [comment]
yes, perhaps to avoid bias in a fixed data splitting in the multi gpu case

**Comment by behxyz** [comment]
Right, especially because in that use case the the patches are extracted in a sequence from tumor or normal region, so although the whole batch is balances any sub-batch is highly imbalanced.

**Comment by wyli** [comment]
I guess to do it properly there should be a buffering stage before model forwarding to allow for sufficient shuffling...

**Comment by behxyz** [comment]
> ```python
> trainer = ...
> 
> @trainer.on(Events.ITERATION_STARTED(custom_event_filtering))
> def switch_batch(engine):
>     engine.state.batch = preprocess_batch(engine.state.batch)
> ```

@vfdev-5, @wyli 
In MONAI SupervisedTrainer:
https://github.com/Project-MONAI/MONAI/blob/8d70890a4b39486109912dfdd9a03e198c7c4736/monai/engines/trainer.py#L147
we already apply a `prepare_batch` function, so `engine.state.batch` is already a processed batch and the type is tuple here, right? or the handler is being triggered before this line?

**Comment by behxyz** [comment]
> I guess to do it properly there should be a patch buffering stage before model.forward to allow for sufficient shuffling...

@wyli, could you elaborate on that?

**Comment by wyli** [comment]
> > I guess to do it properly there should be a patch buffering stage before model.forward to allow for sufficient shuffling...
> 
> 
> 
> @wyli, could you elaborate on that?

yes, even if we shuffle at the image level, the patch level batch might be biased toward a specific subject at each iteration. we can construct a patch buffer and shuffle the buffer to mix patches from different images

**Comment by behxyz** [comment]
> > > I guess to do it properly there should be a patch buffering stage before model.forward to allow for sufficient shuffling...
> > 
> > 
> > @wyli, could you elaborate on that?
> 
> yes, even if we shuffle at the image level, the patch level batch might be biased toward a specific subject at each iteration. we can construct a patch buffer and shuffle the buffer to mix patches from different images

Oh right! similar to having a shuffling buffer before batching images in other use cases.

**Comment by Nic-Ma** [comment]
@wyli @behxyz , I think we are aligned on the direction of this PR now?
Thanks.

**Comment by behxyz** [comment]
I have communicated with other researchers about this and they confirmed we already discussed. It shouldn't have any effect on single-gpu training but it is required for multi-gpu training because since the model has BN layers. We are going to explore replacing BN layers by SyncBN layers for our pathology use case but it requires tweaking any models that is being used.

However, in general, I believe shuffling batches is needed in MONAI for people who are using the torchvision models on multi-gpu with any systematic order of data in their batch. What do you think? @wyli @vfdev-5

**Comment by vfdev-5** [comment]
@behxyz i'd say this could be helpful in ignite too :)

**Comment by ericspod** [comment]
For randomized transforms we use [Randomizable](https://github.com/Project-MONAI/MONAI/blob/af1ffd60d377fa2fc3cf8c43e8057b1ac52559b4/monai/transforms/compose.py#L76) to provide a mechanism for setting the random state object. This is Numpy rather than Pytorch but I think the intent makes sense, we should have a way to set the random state for `PrepareBatchShuffle` or at least its seed separately from a global state as set with `set_determinism`. Should we also have an example and a test of using these with an actual Trainer object?

**Comment by wyli** [comment]
> I guess to do it properly there should be a patch buffering stage before model.forward to allow for sufficient shuffling...

pytorch 1.8 comes with a `BufferedShuffleDataset` https://github.com/pytorch/pytorch/blob/v1.8.0/torch/utils/data/dataset.py#L257 perhaps could resolve this

**Comment by behxyz** [comment]
> > I guess to do it properly there should be a patch buffering stage before model.forward to allow for sufficient shuffling...
> 
> pytorch 1.8 comes with a `BufferedShuffleDataset` https://github.com/pytorch/pytorch/blob/v1.8.0/torch/utils/data/dataset.py#L257 perhaps could resolve this

Thank you @wyli! I'll take a look into it.
